src/helpers.rs

```diff
@@ -857,10 +857,28 @@ pub struct CurrentSpan<'a, 'mir, 'tcx> {
 }
 
 impl<'a, 'mir, 'tcx> CurrentSpan<'a, 'mir, 'tcx> {
+    /// Get the current span, skipping non-local frames.
+    /// This function is backed by a cache, and can be assumed to be very fast.
     pub fn get(&mut self) -> Span {
         *self.span.get_or_insert_with(|| Self::current_span(self.machine))
     }
 
+    /// Similar to `CurrentSpan::get`, but retrieves the parent frame of the first non-local frame.
+    /// This is useful when we are processing something which occurs on function-entry and we want
+    /// to point at the call to the function, not the function definition generally.
+    #[inline(never)]
+    pub fn get_parent(&mut self) -> Span {
+        let stack = self.machine.threads.active_thread_stack();
+        for i in (1..stack.len()).rev() {
+            let def_id = stack[i].instance.def_id();
+            if def_id.is_local() || self.machine.local_crates.contains(&def_id.krate) {
```

Comment:
Btw, should these take into account `track_caller` attributes the way our backtraces do?

Reply:
Probably. This looks tricky to me because we would need access to the `TyCtxt` but after here, we only have access to the `Evaluator` due to the mutable borrow of all of `this`:
https://github.com/rust-lang/miri/blob/df3c141762b4082ef405bf092b385ca859f1dc8b/src/stacked_borrows/mod.rs#L929-L939

Reply:
`TyCtxt` is copy so we can just grab a a copy of taht before the mutable borrow.